### ISSUE TITLE
Training XSeg doesn't continue, doesn't error.

### ISSUE THREAD

**Issue by jarviscodes** [iperov/DeepFaceLab#5202]
## Expected behavior

After going through extraction and basic XSeg editing, I'm trying to run the XSeg_train.bat. 
The first iteration kicks off pretty fast. 
Expected: Continue training after first step or throw an error.

## Actual behavior

Script hangs indefinitely (let it run for 12-14 hours) after output below:

```
Running trainer.


Model first run.

Choose one or several GPU idxs (separated by comma).

[CPU] : CPU
  [0] : GeForce GTX 1080

[0] Which GPU indexes to choose? : 0
0

[wf] Face type ( h/mf/f/wf/head ?:help ) :
wf
[4] Batch_size ( 2-16 ?:help ) :
4
Loading samples: 100%|############################################################| 2179/2179 [00:11<00:00, 193.29it/s]
Loading samples: 100%|############################################################| 1319/1319 [00:07<00:00, 178.98it/s]
Filtering: 100%|##################################################################| 3498/3498 [00:03<00:00, 961.40it/s]
Using 188 segmented samples.
============= Model Summary =============
==                                     ==
==        Model name: XSeg             ==
==                                     ==
== Current iteration: 0                ==
==                                     ==
==----------- Model Options -----------==
==                                     ==
==         face_type: wf               ==
==        batch_size: 4                ==
==                                     ==
==------------ Running On -------------==
==                                     ==
==      Device index: 0                ==
==              Name: GeForce GTX 1080 ==
==              VRAM: 8.00GB           ==
==                                     ==
=========================================
Starting. Press "Enter" to stop training and save model.

Trying to do the first iteration. If an error occurs, reduce the model parameters.
```

The first time I tried this I ran it with batch_size 16, next 8, now at 4. Have not tried lower, but 12 hours for no additional iterations seems rather long.

Note: The preview for the first iteration is shown correctly. I can cycle through the different loaded masks for that iteration.

## Steps to reproduce

run `5.XSeg) train.bat` after extracting and aligning faces, and editing masks.

## Other relevant information

Specs:
* Intel i7-6700K (4GHz)
* 32GB RAM (Already increased pagefile on SSD to 60 GB)
* 64 bit 
* Windows 10 V 1909 Build 18363.1256
* Python Version: The one that came with a fresh DFL Download yesterday.
* Driver version: 460.89
* Other Cuda installations: Removed entirely
* DFL on OS Boot disk? Yes, but a different partition. 

I also notice specifically low usage, when usually training models my GPU Memory (8GB) is used up to 6.8GB, DFL is not taking more than 4.8GB VideoRAM. Also only using 1.4GB of RAM with more than 28 available.
My CPU is near idle, and so is my GPU. See imgur ss below.

![Screenshot](https://i.imgur.com/u88ikE4.png)

Please let me know if there's more debugging logs I can attach.

**Comment by jarviscodes** [comment]
Additional print-screen to verify python version:
![Print screen](https://i.imgur.com/9CsZ3J6.png)

**Comment by jarviscodes** [comment]
* ~~DFL on OS Boot disk? Yes, but a different partition.~~ I have now moved DFL to the Boot partition, the behavior remains the same.

I also turned to modifying `Model.py` in the `Model_XSeg` folder of this release.
I've added the following print statements to the `train` function on line 112.

```python
            def train(input_np, target_np):
                print("IN TRAIN!!")
                l, _ = nn.tf_sess.run ( [loss, loss_gv_op], feed_dict={self.model.input_t :input_np, self.model.target_t :target_np })
                print("POST tf_sess.run!")
                return l
            self.train = train
```

The output is now the following:

```
IN TRAIN!!
POST tf_sess.run!
IN TRAIN!!
```

It appears the second instance of `tf_sess.run()` is failing to complete.

**Comment by jarviscodes** [comment]
SAEHD Training does run smoothly. I'll just give that a go for now, but I'm guessing it doesn't use / need my masks then. 

If you need more information regarding this (XSeg Train) issue, feel free to ask.

**Comment by jarviscodes** [comment]
https://github.com/iperov/DeepFaceLab/commit/554217d0265a35b8570605a3fe57b5be7e06275e

Turning on this setting has indeed fixed things. If you have an NVidia driver > 450.99 and you do not have this option, make sure you are on at least Windows version 2004.

**Comment by Micron-99** [comment]
I'm having a similar issue.  I'm trying to run XSeg Train on Paperspace using a CUDA P4000.
I do not believe the gpu supports hardware accelerated gpu scheduling so it's not even an option.

When I run XSeg Train - it does one interation only.  It's still running but no matter how long I wait it won't perform more iterations.

**Comment by jarviscodes** [comment]
@Micron-99 is that on one of their Windows Server 2019 instances?
I know Windows Server supports HAGS since 1809 version, and have seen docker nodes where it's active on Windows Server.

Regarding P4000, I would be surprised if Nvidia did not create a WDDM2.7 compatible driver for the quadro series.

I think your best bet is to contact paperspace support in order to enable HAGS, or wait for a stable solution from tensorflow... since it's not necessarily a DFL Problem but is more something between Tensorflow and Windows.
For some reason `tf_sess.run()` just hangs completely with HAGS not enabled on complex graphs.

You could also follow [this issue](https://github.com/tensorflow/tensorflow/issues/45648) iperov has opened at tensorflow that describes the same behavior.

**Comment by BigSegun** [comment]
> ## Expected behavior
> After going through extraction and basic XSeg editing, I'm trying to run the XSeg_train.bat. The first iteration kicks off pretty fast. Expected: Continue training after first step or throw an error.
> 
> ## Actual behavior
> Script hangs indefinitely (let it run for 12-14 hours) after output below:
> 
> ```
> Running trainer.
> 
> 
> Model first run.
> 
> Choose one or several GPU idxs (separated by comma).
> 
> [CPU] : CPU
>   [0] : GeForce GTX 1080
> 
> [0] Which GPU indexes to choose? : 0
> 0
> 
> [wf] Face type ( h/mf/f/wf/head ?:help ) :
> wf
> [4] Batch_size ( 2-16 ?:help ) :
> 4
> Loading samples: 100%|############################################################| 2179/2179 [00:11<00:00, 193.29it/s]
> Loading samples: 100%|############################################################| 1319/1319 [00:07<00:00, 178.98it/s]
> Filtering: 100%|##################################################################| 3498/3498 [00:03<00:00, 961.40it/s]
> Using 188 segmented samples.
> ============= Model Summary =============
> ==                                     ==
> ==        Model name: XSeg             ==
> ==                                     ==
> == Current iteration: 0                ==
> ==                                     ==
> ==----------- Model Options -----------==
> ==                                     ==
> ==         face_type: wf               ==
> ==        batch_size: 4                ==
> ==                                     ==
> ==------------ Running On -------------==
> ==                                     ==
> ==      Device index: 0                ==
> ==              Name: GeForce GTX 1080 ==
> ==              VRAM: 8.00GB           ==
> ==                                     ==
> =========================================
> Starting. Press "Enter" to stop training and save model.
> 
> Trying to do the first iteration. If an error occurs, reduce the model parameters.
> ```
> 
> The first time I tried this I ran it with batch_size 16, next 8, now at 4. Have not tried lower, but 12 hours for no additional iterations seems rather long.
> 
> Note: The preview for the first iteration is shown correctly. I can cycle through the different loaded masks for that iteration.
> 
> ## Steps to reproduce
> run `5.XSeg) train.bat` after extracting and aligning faces, and editing masks.
> 
> ## Other relevant information
> Specs:
> 
> * Intel i7-6700K (4GHz)
> * 32GB RAM (Already increased pagefile on SSD to 60 GB)
> * 64 bit
> * Windows 10 V 1909 Build 18363.1256
> * Python Version: The one that came with a fresh DFL Download yesterday.
> * Driver version: 460.89
> * Other Cuda installations: Removed entirely
> * DFL on OS Boot disk? Yes, but a different partition.
> 
> I also notice specifically low usage, when usually training models my GPU Memory (8GB) is used up to 6.8GB, DFL is not taking more than 4.8GB VideoRAM. Also only using 1.4GB of RAM with more than 28 available. My CPU is near idle, and so is my GPU. See imgur ss below.
> 
> ![Screenshot](https://camo.githubusercontent.com/911d10b96467edd832ab3d0cf0ff98a0bfbe46c00203622a836e7a499313ed90/68747470733a2f2f692e696d6775722e636f6d2f753838696b45342e706e67)
> 
> Please let me know if there's more debugging logs I can attach.

HELLO BROTHER I HAVE AN ISSUE WHEN RUNNING MY XSEG TRAINER ONCE I GET TO (ENABLE PRETRAINING MODE)
IT GIVES ME A LONG ERROR AND IT WILL NOT PASS THERE 
IT WILL NOT START FILTERING

**Comment by BigSegun** [comment]
![0df540f5-4563-4bea-88b6-6544bda82a61](https://user-images.githubusercontent.com/68261548/230757662-bf702945-5e45-46f5-8420-fce38cb4b983.jpeg)


this is the picture of the error
![0df540f5-4563-4bea-88b6-6544bda82a61](https://user-images.githubusercontent.com/68261548/230757699-c1f50e59-657f-430c-a809-7263030e9303.jpeg)

**Comment by jarviscodes** [comment]
@BigSegun This does not seem related to the original issue, you actually have an allocation error.

Without the proper HAGS setting, the training just hung and didn't throw a `MemoryError` it might be that this behavior is now different, but if you have validated that HAGS setting and this issue is still occuring, you're going to want to tweek your training settings to use less VRAM.

Also, this topic is closed, so if you have more issues, you might want to open a new issue with detailed logs and steps to reproduce, including configs and hardware setup.
